sdks/java/extensions/sketching/src/main/java/org/apache/beam/sdk/extensions/sketching/MostFrequent.java

```diff
@@ -0,0 +1,747 @@
+          for (KV<ElementWrapper<V>, Long> pair : kv.getValue().estimateTopK(k)) {
+            c.output(KV.of(
+                kv.getKey(),
+                KV.of(pair.getKey().getElement(), pair.getValue())));
+          }
+        }
+      };
+    }
+
+    /**
+     * {@link DoFn} that retrieves the top k most frequent elements per key in a
+     * {@link PCollection} of {@link KV}s with {@link StreamSummarySketch} containing
+     * serializable objects.
+     *
+     * @param k     the top k most frequent elements to return
+     */
+    public static <K, V> DoFn<KV<K, StreamSummarySketch<V>>,
+        KV<K, KV<V, Long>>> perKeySerializable(int k) {
+      return new DoFn<KV<K, StreamSummarySketch<V>>, KV<K, KV<V, Long>>>() {
+        @ProcessElement
+        public void processElement(ProcessContext c) {
+          KV<K, StreamSummarySketch<V>> kv = c.element();
+          for (KV<V, Long> pair : kv.getValue().estimateTopK(k)) {
+            c.output(KV.of(
+                kv.getKey(),
+                KV.of(pair.getKey(), pair.getValue())));
+          }
+        }
+      };
+    }
+  }
+
+  /** Wrap StreamLib's Stream Summary sketch. */
+  public static class StreamSummarySketch<T> extends StreamSummary<T> {
+
+    public StreamSummarySketch() {
+      super();
+    }
+
+    public StreamSummarySketch(int capacity) {
+      super(capacity);
+    }
+
+    /** Merge several {@link StreamSummarySketch} into this one. */
+    public void merge(StreamSummarySketch<T>... sketches) {
+      for (StreamSummarySketch<T> sketch : sketches) {
+        List<Counter<T>> top = sketch.topK(sketch.size());
+        for (Counter<T> counter : top) {
+          offer(counter.getItem(), (int) counter.getCount());
+        }
+      }
+    }
+
+    /**
+     * Retrieves the {@code k} most frequent elements in a {@link StreamSummarySketch}
+     * sketch and returns a list of {@link KV}s where the pair at index i represents
+     * the ith most frequent element associated with its estimate count.
+     *
+     * @param k     the top k most frequent elements to return
+     */
+    public List<KV<T, Long>> estimateTopK(int k) {
+      List<Counter<T>> top = super.topK(k);
+      List<KV<T, Long>> output = new ArrayList<>(k);
+      for (Counter<T> c : top) {
+        output.add(KV.of(c.getItem(), c.getCount()));
+      }
+      return output;
+    }
+
+    @Override
+    public int hashCode() {
+      return this.toString().hashCode();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (this == o) {
+        return true;
+      } else if (!(o instanceof StreamSummarySketch)) {
+        return false;
+      }
+      StreamSummarySketch that = (StreamSummarySketch) o;
+      return this.toString().equals(that.toString());
+    }
+  }
+
+  /**
+   * This class is used to wrap a non serializable type into a serializable wrapper which
+   * uses the {@link Coder} as serializer. This wrapper is automatically used when calling
```

Comment:
Why do we need this wrapper? If it's for the sake of making the StreamSummary itself encodable, I think you could get around that by encoding it using its .peek(capacity) and decoding symmetrically.

Reply:
It's not to make it encodage but to make it consistent with equals(), due to the Illegal mutation error I got

Reply:
This you can circumvent by creating a wrapper around StreamSummary with a properly defined equals() / hashCode().

Reply:
Whoops I thought you were talking about the StreamSummarySketch wrapper.

This wrapper is used for non serizaliable elements because StreamSummary calls writeObject() for tracked elements therefore when elements are not serializable this is a problem.

But this problem will be solved if we encode the StreamSummary as you said above.